### PR TITLE
Add csd connectivity

### PR DESCRIPTION
**Pull request by kajal5888**
Author Guidelines
-----------------
- [x] Is the change set **< 600 lines**?
- [ ] Was the code checked for memory leaks/performance bottlenecks?
- [X] Is the code running locally **and** on the ESI cluster?
- [ ] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [X] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [X] Are all docstrings complete and accurate?
- [ ] Is the CHANGELOG.md up to date?

### CODE REVIEW

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 1]
```diff
@@ -378,33 +380,35 @@ def connectivityanalysis(data, method="coh", keeptrials=False, output="abs",
...
-                           log_dict=log_dict)
-
-    # attach potential older cfg's from the input
-    # to support chained frontend calls..
-    out.cfg.update(data.cfg)
-    # attach frontend parameters for replay
-    out.cfg.update({'connectivityanalysis': new_cfg})
-    return out
+    if method == 'csd':
+        new_cfg.update({'output': st_out.data.dtype.name})
```
Comment (tensionhead) [inline syncopy/nwanalysis/connectivity_analysis.py:384]: What we want to save in the `cfg` are parameter values which can be re-used for the same frontend call, what you are saving here won't work, `output` has to be one of `connectivity_outputs` and already gets attached correctly here:
https://github.com/esi-neuroscience/syncopy/blob/83455e70a94369c900e1688033c0766ac6565751/syncopy/nwanalysis/connectivity_analysis.py#L219
So this line can be removed.

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 2]
```diff
@@ -378,33 +380,35 @@ def connectivityanalysis(data, method="coh", keeptrials=False, output="abs",
...
+        st_out.cfg.update(data.cfg)
+        st_out.cfg.update({'cross_spectral': new_cfg})
```
Comment (tensionhead) [inline syncopy/nwanalysis/connectivity_analysis.py:386]: It's very important that the top-level keys inside a `cfg` are frontend meta function names, like `connectivityanalysis` or `freqanalysis`. This is how the `unwrap_cfg` decorator knows which part of a `cfg` needs to be applied to the current frontend call.

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 3]
```diff
@@ -378,33 +380,35 @@ def connectivityanalysis(data, method="coh", keeptrials=False, output="abs",
...
-    av_compRoutine.compute(st_out, out, parallel=kwargs.get("parallel"),
-                           log_dict=log_dict)
-
-    # attach potential older cfg's from the input
-    # to support chained frontend calls..
-    out.cfg.update(data.cfg)
-    # attach frontend parameters for replay
-    out.cfg.update({'connectivityanalysis': new_cfg})
-    return out
+    if method == 'csd':
```
Comment (tensionhead) [inline syncopy/nwanalysis/connectivity_analysis.py:383]: nice and clean solution with a simple `if/else` :+1:

**Review by tensionhead — COMMENTED** [review]
Nice work @kajal5888 ! See my detailed comments below, I think most critical would be to rename the `new_cfg` entry for this frontend, as this right now breaks replayability. 

You would also see this in a proper test however, would be great you could add a  `class TestCSD` in `/tests/test_connectivity.py`. The easiest test I could think of, is to take the same synth data as in `TestCoherence` , and plug the new output with `method='csd'` into an existing csd normalization routine to recover the coherence from that csd.

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 4]
```diff
@@ -4,32 +4,30 @@
...
+from syncopy.shared.tools import get_defaults, best_match, get_frontend_cfg
+from syncopy.shared.parsers import data_parser, scalar_parser
+import numpy as np
+# Syncopy imports
 
 
-availableMethods = ("coh", "corr", "granger")
-coh_outputs = {"abs", "pow", "complex", "fourier", "angle", "real", "imag"}
+availableMethods = ("coh", "corr", "granger", "csd")
+connectivity_outputs = {"abs", "pow", "complex", "fourier", "angle", "real", "imag"}
```
Comment (tensionhead) [inline syncopy/nwanalysis/connectivity_analysis.py:30]: I am ok with the renaming, but then you also have to change the name in the respective `TestCoherence`, this is why the pipeline fails atm :wink:

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 5]
```diff
@@ -4,32 +4,30 @@
...
+from syncopy.shared.kwarg_decorators import (unwrap_cfg, unwrap_select,
+                                             detect_parallel_client)
+from syncopy.shared.errors import (
+    SPYValueError,
+    SPYWarning,
+    SPYInfo)
+from syncopy.datatype import CrossSpectralData, AnalogData, SpectralData
+from syncopy.shared.tools import get_defaults, best_match, get_frontend_cfg
+from syncopy.shared.parsers import data_parser, scalar_parser
+import numpy as np
```
[the remark this answers is not in the document]
Comment (kajal5888) [reply syncopy/nwanalysis/connectivity_analysis.py:25]: formated using pep8

[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/nwanalysis/connectivity_analysis.py:25]: mmh, did not know there is a pep8 standard for the order of imports?! Maybe you auto-formatter just made this decision :man_shrugging:

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 6]
```diff
@@ -72,6 +69,18 @@ def connectivityanalysis(data, method="coh", keeptrials=False, output="abs",
         * **nTaper** : (optional) number of orthogonal tapers for slepian tapers
         * **pad**: either pad to an absolute length in seconds or set to `'nextpow2'`
 
+    "csd" : ('Multi-) tapered cross spectral density estimate
+        Computes the normalized cross spectral densities between all channel combinations
```
[the remark this answers is not in the document]
Comment (kajal5888) [reply syncopy/nwanalysis/connectivity_analysis.py]: updated

File: syncopy/nwanalysis/connectivity_analysis.py (hunk 2 above)
Comment (kajal5888) [reply syncopy/nwanalysis/connectivity_analysis.py]: done

File: syncopy/nwanalysis/connectivity_analysis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kajal5888) [reply syncopy/nwanalysis/connectivity_analysis.py:25]: I did the formatting using pep8

[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/nwanalysis/connectivity_analysis.py:25]: yes, that you already said :smile: still that does not move our imports back where they belong ;)

**Review by tensionhead — CHANGES_REQUESTED** [review]

File: syncopy/tests/test_connectivity.py [hunk 7]
```diff
@@ -648,25 +693,27 @@ def test_corr_cfg(self):
...
+    def test_parallel(self):
+        check_parallel(TestCorrelation())
 
     def test_corr_polyremoval(self):
 
         call = lambda polyremoval: self.test_corr_solution(polyremoval=polyremoval)
         helpers.run_polyremoval_test(call)
 
 
+def check_parallel(TestClass, testcluster=None):
```
Comment (tensionhead) [inline syncopy/tests/test_connectivity.py:705]: This makes a lot of sense :+1: You could even move this function to `/tests/helpers.py` ;)

File: syncopy/tests/test_connectivity.py [hunk 8]
```diff
@@ -516,6 +505,62 @@ def test_coh_outputs(self):
...
+        check_parallel(TestCSD())
+
+    def test_csd_input(self):
+        assert isinstance(self.spec, SpectralData)
+
+    def test_csd_cfg(self):
+        Method = 'csd'
+        cross_spec = spy.connectivityanalysis(self.spec, method=Method)
+        assert len(cross_spec.cfg) == 2
+        assert np.all([True for cfg in zip(self.spec.cfg['freqanalysis'], cross_spec.cfg['freqanalysis']) if cfg[0] == cfg[1]])
```
[the remark this answers is not in the document]
Comment (kajal5888) [reply syncopy/tests/test_connectivity.py:560]: done!

**Review by tensionhead — APPROVED** [review]
Looks good, thx @kajal5888 !

**Pull request merged**